### PR TITLE
fix(sdk): only apply importer node when compiling to IR

### PR DESCRIPTION
**Pull request by chensun**
**Description of your changes:**
Fixes: #5263

The logic for inserting importer node heavily relies on inputs being properly type annotated -- importer node is for input artifact not connected to an upstream component while if an input is not type annotated or annotated using an arbitrary custom string, it's viewed as an artifact in v2. But such inputs in v1 could be parameters and importer logic shouldn't apply.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

**Comment by Ark-kun** [comment]
How many compilation modes do we have? Just v1 and v2? Do we even need to call `_attach_v2_specs` for v1?

File: sdk/python/kfp/dsl/importer_node.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ def build_importer_spec(
   Returns:
     An importer spec.
   """
-  assert bool(pipeline_param_name) != bool(constant_value), (
+  assert bool(pipeline_param_name == None) != bool(constant_value == None), (
```
Comment (Ark-kun) [inline sdk/python/kfp/dsl/importer_node.py:43]: A more correct testing for Non is `pipeline_param_name is None`.
So
```
assert (pipeline_param_name is None) != (constant_value is None), (
```
Another variant might be:
```
assert not(pipeline_param_name is not None and constant_value is not None)
assert not(pipeline_param_name is None and constant_value is None)

```

**Comment by Ark-kun** [comment]
/lgtm
/hold
Please unhold as needed.

File: sdk/python/kfp/dsl/importer_node.py (hunk 1 above)
Comment (chensun) [reply sdk/python/kfp/dsl/importer_node.py]: Done.

**Comment by chensun** [comment]
> How many compilation modes do we have? Just v1 and v2? Do we even need to call `_attach_v2_specs` for v1?

There're three modes introduced by https://github.com/kubeflow/pipelines/pull/5218:
https://github.com/kubeflow/pipelines/blob/ea3d22fa659ca16abdc6b871811fe99a7fa8db71/sdk/python/kfp/dsl/_pipeline.py#L32-L40

`V1_LEGACY` and `V2_COMPATIBLE` both use v1 compiler, and we need to call `_attach_v2_specs` for `V2_COMPATIBLE` mode but not `V1_LEGACY` mode.
The flag shown in this PR `is_compiling_for_v2` isn't integrated with any of these modes yet as they weren't available by the time we merge the dsl. It's currently based on the callstack -- whether v2 compiler method appears in the callstack:

https://github.com/kubeflow/pipelines/blob/ea3d22fa659ca16abdc6b871811fe99a7fa8db71/sdk/python/kfp/dsl/_component_bridge.py#L172-L176

I'll make a separate to make `is_compiling_for_v2` based on compiler mode setting. Will limit `_attach_v2_specs` call in that PR.

**Comment by neuromage** [comment]
I think we need to rethink the auto-insertion of Importer. Can this be done on request? E.g. a pipeline which annotates some input parameter as a URI to be imported:

```
def my_pipeline(input_dataset_uri: ImporterUri, ...):

  my_preprocessing_op = MyOp(input_uri=input_dataset_uri, ....)

```

So we'd introduce the type `ImporterUri` to annotate strings that refer to Artifact uris to be imported. This makes the behaviour more explicit. WDYT?

**Comment by chensun** [comment]
> I think we need to rethink the auto-insertion of Importer. Can this be done on request? E.g. a pipeline which annotates some input parameter as a URI to be imported:
> 
> ```
> def my_pipeline(input_dataset_uri: ImporterUri, ...):
> 
>   my_preprocessing_op = MyOp(input_uri=input_dataset_uri, ....)
> ```
> 
> So we'd introduce the type `ImporterUri` to annotate strings that refer to Artifact uris to be imported. This makes the behaviour more explicit. WDYT?

We basically do what you described here in v2 expect that importer is tied to `InputUri` instead of `ImporterUri`. 

For v1 though, we have to disable this error, otherwise it could be hit a lot due to inputs mis-categorized as artifacts.
https://github.com/kubeflow/pipelines/blob/0c933971352da393bf79325249d031407da3409f/sdk/python/kfp/dsl/_component_bridge.py#L215-L223

So my thought was since importer isn't supported by v1 engine at this moment, we should just skip it completely in v1 until we have a clear picture on how v1 could support importer. WDYT?

**Comment by neuromage** [comment]
/lgtm

Yes, this PR definitely makes sense. But let's discuss syntax for v2 importer offline.

**Comment by Ark-kun** [comment]
> There're three modes introduced by #5218:
>  
> `V1_LEGACY` and `V2_COMPATIBLE` both use v1 compiler, and we need to call `_attach_v2_specs` for `V2_COMPATIBLE` mode but not `V1_LEGACY` mode.
> The flag shown in this PR `is_compiling_for_v2` isn't integrated with any of these modes yet as they weren't available by the time we merge the dsl. It's currently based on the callstack -- whether v2 compiler method appears in the callstack:

Thank you for the explanation.
/lgtm

>I think we need to rethink the auto-insertion of Importer. Can this be done on request? E.g. a pipeline which annotates some input parameter as a URI to be imported:

+1

>For v1 though, we have to disable this error, otherwise it could be hit a lot due to inputs mis-categorized as artifacts.

I think that in the case of that error the we can insert an "uploader" step, which takes primitive parameter and outputs an artifact.

**Pull request merged**